core/src/main/java/com/linecorp/armeria/internal/server/annotation/AnnotatedService.java

```diff
@@ -251,7 +251,30 @@ Route route() {
 
     @Override
     public HttpResponse serve(ServiceRequestContext ctx, HttpRequest req) throws Exception {
-        final HttpResponse response = HttpResponse.from(serve0(ctx, req));
+        if (!defaultHttpHeaders.isEmpty()) {
+            ctx.mutateAdditionalResponseHeaders(mutator -> mutator.add(defaultHttpHeaders));
+        }
+        if (!defaultHttpTrailers.isEmpty()) {
+            ctx.mutateAdditionalResponseTrailers(mutator -> mutator.add(defaultHttpTrailers));
+        }
+
+        final AggregationType aggregationType =
+                AnnotatedValueResolver.aggregationType(aggregationStrategy, req.headers());
+
+        final HttpResponse response;
+        if (aggregationType == AggregationType.NONE && !useBlockingTaskExecutor) {
+            // Fast-path: No aggregation required and blocking task executor is not used.
+            final Object result = invoke(ctx, req, AggregatedResult.EMPTY);
+            if (responseType == ResponseType.HTTP_RESPONSE) {
+                assert result != null;
+                response = (HttpResponse) result;
+            } else {
+                response = convertResponse(ctx, result);
```

Comment:
It seems like scala futures aren't being converted correctly. I'm guessing the cause is a missing `toCompletionStage` call 😅  (or at least the following test fails in my local)
Do you mind checking?

```
diff --git a/scala/scala_2.13/src/test/scala/com/linecorp/armeria/client/scala/RestClientSuite.scala b/scala/scala_2.13/src/test/scala/com/linecorp/armeria/client/scala/RestClientSuite.scala
index a9ce10442..393a06501 100644
--- a/scala/scala_2.13/src/test/scala/com/linecorp/armeria/client/scala/RestClientSuite.scala
+++ b/scala/scala_2.13/src/test/scala/com/linecorp/armeria/client/scala/RestClientSuite.scala
@@ -42,6 +42,19 @@ class RestClientSuite extends FunSuite with ServerSuite {
       },
       Array.emptyObjectArray: _*
     )
+    sb.annotatedService(
+      new {
+        @Get
+        @Post
+        @Put
+        @Delete
+        @Patch
+        @ProducesJson
+        @Path("/future")
+        def future(): Future[RestResponse] = Future.successful(RestResponse("1", "Hi!"))
+      },
+      Array.emptyObjectArray: _*
+    )
 
     sb.annotatedService(
       new {
@@ -117,6 +130,18 @@ class RestClientSuite extends FunSuite with ServerSuite {
     assertEquals(content.content, "content")
   }
 
+  test("future output") {
+    val restClient = server.webClient().asScalaRestClient()
+    val future =
+      restClient
+        .post("/future")
+        .execute[RestResponse]()
+
+    val content = Await.result(future, Duration.Inf).content()
+    assertEquals(content.id, "1")
+    assertEquals(content.content, "Hi!")
+  }
+
   test("ScalaRestClientPreparation should return self type") {
     ReflectionUtils
       .getMethods(classOf[RequestPreparationSetters])
hannam@AL01325349 armeria % 

```

Reply:
Can we just call
`response = convertResponse(ctx, invoke(ctx, req, AggregatedResult.EMPTY));`
because `convertResponse` returns right away if the `result` is an instanceof `HttpResponse`?